KubeOne: documentation for Operating System Manager

ahmedwaleedmalik opened the pull request:
Signed-off-by: Waleed Malik <ahmedwaleedmalik@gmail.com>

**Special notes for your reviewer**:
The diagrams have not been updated yet. I've intentionally requested reviews from multiple people because this documentation is super critical for us and our customers. Looking forward to the feedback ;)

toschneck commented:
In general I like the created content. But I think a few things are potential to challange

- Doc location: In my opinion `Guides` is not the right location, it should be more under `Architecture` as it is describing more the concept.
- Missing deeper content how to realy work with it. Somehow I miss:
  - Reference to the CRDs (what could be done or not)
  - Where do I find the defaults
  - How could I customize it? 
  - Examples for customization, e.g. for the offline setup @hc2p  can potential share an example

hc2p commented on content/kubeone/master/guides/operating-system-manager/_index.en.md:
```diff
@@ -0,0 +1,50 @@
+
+Immutable resource that contains the **actual configurations** that are going to be used to bootstrap and provision the worker nodes. It is a subset of OperatingSystemProfile, OperatingSystemProfile is a template while OperatingSystemConfig is an instance rendered using OperatingSystemProfile, MachineDeployment and flags provided at OSM level.
+
+OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping and provisioning secrets to become available. Once they are ready, it will extract the configurations from those se ...
+to the machines that it's going to provision.
```
```suggestion
OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping- and provisioning-secrets to become available. Once they are ready, it will extract the configurations from those secrets and pass them as `user-data` to the to-be-provisioned machines.
```

kron4eg commented:
@toschneck sounds reasonable, but items you've listed are on their own can (and probably should) be a dedicated guides :)

hc2p commented on content/kubeone/master/guides/operating-system-manager/_index.en.md:
```diff
@@ -0,0 +1,50 @@
+
+For each MachineDeployment we have two types of configurations, which are stored in secrets:
+
+1. **Bootstrap**: Configuration ussed for initially setting up the machine and to fetch the provisioning configuration.
```
```suggestion
1. **Bootstrap**: Configuration used for initially setting up the machine and fetching the provisioning configuration.
```

hc2p commented on content/kubeone/master/guides/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+It can be used as follows:
+
+1. Create a custom OSP and place it in a folder inside your custom addons path, if it exists otherwise create one. Let's take this minimalistic OSP just for the sake of understanding:
```
```suggestion
1. Create a custom OSP and place it in a folder inside your custom addons path, if it exists, otherwise create one. Let's take this minimalistic OSP as an example:
```

hc2p commented on content/kubeone/master/guides/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+## Additional Information
+
+- If an `OperatingSystemProfile` is updated, that wouldn't result in an automatic rotation of the machines. This is an intentional design decision since an `OperatingSystemProfile` can be associated with multiple Machine Deployments and it is the user's responsibility to rotate the machines when the `OperatingSystemProfile` is updated.
```
```suggestion
- An update to `OperatingSystemProfile` will not result in an automatic rotation of the machines. This is an intentional design decision since an `OperatingSystemProfile` can be associated with multiple Machine Deployments. It is the user's responsibility to rotate the machines when the `OperatingSystemProfile` is updated.
```

hc2p reviewed (commented):
Just found some typos and potential wording improvements. 

Wondering where to add the airgap example which I'm happy to provide. Maybe we could do something similar like [the cloud-init docs](https://cloudinit.readthedocs.io/en/latest/topics/examples.html) and add an example page?

ahmedwaleedmalik commented:
> Doc location: In my opinion Guides is not the right location, it should be more under Architecture as it is describing more the concept

Agreed and moved it under the Architecture section.

> Missing deeper content how to realy work with it

I agree with @kron4eg, these topics are not KubeOne specific and should be covered in dedicated guides.

> Just found some typos and potential wording improvements.

Thank you, updated the docs.

> Wondering where to add the airgap example which I'm happy to provide.

Once we have the KKP and KubeOne specific docs in place we'll try to add those generic examples as well. Basically, I want to avoid duplication and an idea was floated to have dedicated docs for OSM, just like KubeOne and KKP. But we haven't made a solid decision on that yet.

kron4eg commented:
In general looks good to me, but as @toschneck said, this topic in general has big potential for a lot of content around it. But let's get started at least with something!

/approve

:zap:

xmudrii commented:
/assign
I'll take a look at this mostly like tomorrow.

xmudrii commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
````diff
@@ -0,0 +1,145 @@
+2. Create your infrastructure
+
+```bash
+terraform apply -var=initial_machinedeployment_operating_system_profile=osp-install-curl
````
You should probably put this in `terraform.tfvars` instead.

ahmedwaleedmalik replied:
For demonstration purpose, I don't see any issue with this but sure.

xmudrii commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+**NOTE:** The path for addons can be anything. But make sure that the custom OSP exists in that directory.
+
+### For an existing machine deployments
```
```suggestion
### For an existing cluster
```

xmudrii commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+{{% notice warning %}}
+Secrets and OSC against the Machine Deployment will be rotated right away when we update the annotation. But the existing Machine Deployments will not be rotated automatically.
+Existing machine deployment should be rotated manually, for them to consume configurations generated by OSM.
```
```suggestion
Existing MachineDeployments should be rotated manually for them to consume configurations generated by OSM.
```

xmudrii commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+[rolling-restart]: {{< ref "../../../cheat-sheets/rollout-machinedeployment" >}}
+
+### For new machine deployments
+
+Create a new Machine Deployment and make sure that you add the `k8c.io/operating-system-profile` annotation with the required OSP name.
```
This should be before `For existing MachineDeployments`.

xmudrii reviewed (commented):
> Doc location: In my opinion Guides is not the right location, it should be more under Architecture as it is describing more the concept.

I agree for the first document, but it doesn't feel like a right place for the second (usage) document.

> Missing deeper content how to realy work with it.

I agree with all those points, but also agree with @kron4eg that those should be dedicated documents, probably shared between KubeOne and KKP.

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+It can be used as follows:
+
+1. Create a custom OSP and place it in a folder inside your custom addons path, if it exists, otherwise create one. Let's take this minimalistic OSP as an example:
```
That's irrelevant in the scope of KubeOne. We might cover that in the shared docs for OSM.

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+
+The annotation `k8c.io/operating-system-profile` is used at the Machine Deployment level to select the Operating System Profile that will be used to generate configurations for provisioning the machines.
+
+### When creating a new cluster
```
Does machine controller work with bare metal setups? 

> Usually the Kubermatic machine-controller would take care of managing the worker nodes. However, without a provider we can not make use of it. For the following steps, we assume that the required infrastructure is in place and SSH access is ensured.

According to [this](https://docs.kubermatic.com/kubeone/v1.4/tutorials/creating_clusters_baremetal/) doc, it doesn't.

xmudrii replied:
What I thought about is following the document you linked on cloud providers while still using machine-controller. In other words, just opt out from Terraform. I think it's okay as it is right now.

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
```diff
@@ -0,0 +1,145 @@
+For existing Machine Deployments; edit the machine deployment and update `k8c.io/operating-system-profile` annotation with the required OSP name.
+
+{{% notice warning %}}
+Secrets and OSC against the Machine Deployment will be rotated right away when we update the annotation. But the existing Machine Deployments will not be rotated automatically.
```
This is actually correct for existing clusters. MD will be reconciled in OSM only when it has this particular annotation. I'll rephrase it a bit.

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/usage/_index.en.md:
````diff
@@ -0,0 +1,145 @@
+  path: "./custom-addons"
+```
+
+**NOTE:** The path for addons can be anything. But make sure that the custom OSP exists in that directory.
````
> Create a custom OSP and place it in a folder inside your custom addons path, if it exists, otherwise create one

> We should use notice blocks instead of **NOTE:**
Well tbh, they add a lot of noise and in this case they are not worth the visual noise.

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/_index.en.md:
```diff
@@ -0,0 +1,50 @@
+
+For each MachineDeployment we have two types of configurations, which are stored in secrets:
+
+1. **Bootstrap**: Configuration used for initially setting up the machine and fetching the provisioning configuration.
+2. **Provisioning**: Configuration with the actual `cloud-config` that is used to provision the worker machine.
```
Will be explained in another doc

ahmedwaleedmalik commented on content/kubeone/master/architecture/operating-system-manager/compatibility/_index.en.md:
```diff
@@ -0,0 +1,21 @@
++++
+title = "Compatibility Matrix"
```
Sure

ahmedwaleedmalik commented:
> I agree for the first document, but it doesn't feel like a right place for the second (usage) document.

I wouldn't like to scatter the document into multiple sections.

ahmedwaleedmalik commented:
Updated the documentation and pushed changes against comments that have been marked as resolved :)

xmudrii reviewed (approved):
This is a great starting point. I like that both documents are describing the most important parts in an easy-to-understand fashion. Let's get this merged and then we can discuss further improvements and documents. :zap: :shipit: 

/approve

xmudrii commented:
/hold
@ahmedwaleedmalik Leaving it up to you to merge it.

ahmedwaleedmalik commented:
> Let's get this merged and then we can discuss further improvements and documents.

Definitely, as discussed, we have to improve it further and make it even more intuitive for the end user.

/unhold

Removing the hold, let's merge this, and then we'll get back to this in the coming days. I think once I update the documentation for KKP, I'll have a very clear picture of how the shared docs/guides should look like for OSM.
